cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs

```diff
@@ -353,70 +373,137 @@ readSigningKeyFile skFile =
                           AStakePoolSigningKey
       ]
 
-categoriseWitnessSigningKey :: SomeWitnessSigningKey
-                            -> Either (Api.SigningKey Api.ByronKey)
-                                       Api.ShelleyWitnessSigningKey
-categoriseWitnessSigningKey swsk =
+partitionSomeWitnesses
+  :: [ByronOrShelleyWitness]
+  -> ( [Api.SigningKey Api.ByronKey]
+     , [Api.ShelleyWitnessSigningKey]
+     , [Api.MultiSigScript]
+     )
+partitionSomeWitnesses [] = ([], [], [])
+partitionSomeWitnesses (wt' : rest') =
+    go wt' rest' ([], [], [])
+  where
+    go wt [] (bwl, skwl, sswl) =
+      case wt of
+        AByronWitness bw -> (bwl ++ [bw], skwl, sswl)
+        AShelleyKeyWitness skw -> (bwl, skwl ++ [skw], sswl)
+        AShelleyScriptWitness ssw -> (bwl, skwl, sswl ++ [ssw])
+    go wt (next : rest) (bwl, skwl, sswl) =
+      case wt of
+        AByronWitness bw -> go next rest (bwl ++ [bw], skwl, sswl)
+        AShelleyKeyWitness skw -> go next rest (bwl, skwl ++ [skw], sswl)
+        AShelleyScriptWitness ssw -> go next rest (bwl, skwl, sswl ++ [ssw])
```

Comment:
I think this can be simpler. I don't think we need to split on the top level [] vs (:). Currently the go takes a guaranteed non-empty list (wt : rest) as separate wt and rest args. That then forces the unnecessary duplication. But it could just take an ordinary list, and then there's no duplication.

Also note that this is appending to the end, so it is O(n^2), where it could prepend and reverse, giving O(n).

This whole thing can be a single foldl, followed by reversing each element.

Reply:
You're right. I've simplified this in https://github.com/input-output-hk/cardano-node/pull/1871.